src/transformations.rs

```diff
@@ -10,6 +10,111 @@ use std::path::PathBuf;
 
 generate_counter!(FreshVarCounter, usize);
 
+/// # Desugarise destructuring.usize
+///
+/// Replace the let patterns destructuring by the classical let in.
+/// It will first destruct the pattern and create a new var for each field of the patternn.
+/// After that, it will construct a new Record/List from the extracted fields.
+///
+/// ## Example:
+///
+/// Taking the let pattern:
+/// ```text
+/// let x {a,d=b} = {a=1,b=2,c="ignored"} in ...
+/// ```
+/// after transformation we will have:
+/// ```text
+/// let x = (let var = {a=1,b=2,c="ignored"} in {a=var.a,d=var.b}) in (
+///     let a = x.a in (
+///         let d = x.b in ...))
+/// ```
+pub mod desugar_destructuring {
+    use super::{Ident, RichTerm, Term};
+    use crate::destruct::{Destruct, Match};
+    use crate::term::make::{op1, op2};
+    use crate::term::MetaValue;
+    use crate::term::{BinaryOp::DynRemove, UnaryOp::StaticAccess};
+
+    pub fn desugar_with_contract(rt: RichTerm) -> RichTerm {
+        if let Term::LetPattern(x, pat, t_, body) = *rt.term {
+            let pos = body.pos.clone();
+            let meta = pat.clone().as_contract();
+            let t_ = {
+                let t_pos = t_.pos.clone();
+                RichTerm::new(
+                    Term::MetaValue(MetaValue {
+                        value: Some(t_),
+                        ..meta
+                    }),
+                    t_pos,
+                )
+            };
+            desugar(RichTerm::new(Term::LetPattern(x, pat, t_, body), pos))
+        } else {
+            rt
+        }
+    }
+
+    pub fn desugar(rt: RichTerm) -> RichTerm {
+        if let Term::LetPattern(x, pat, t_, body) = *rt.term {
+            let pos = body.pos.clone();
+            let x = if let Some(x) = x {
+                x
+            } else {
+                super::fresh_var()
+            };
+            RichTerm::new(
+                Term::Let(
+                    x.clone(),
+                    t_,
+                    destruct_term(x.clone(), pat.clone(), drop_fields(x, pat, body)),
+                ),
+                pos,
+            )
+        } else {
+            rt
+        }
+    }
+
+    fn drop_fields(x: Ident, pat: Destruct, body: RichTerm) -> RichTerm {
```

Comment:
Just wondering, is it really needed to drop matched fields? Couldn't we just keep the original value around?

Reply:
What do you mean? I actualy gave a bad meaning name to this function. I changed it to `bind_open_field`. Do you think it preferable to pass the `pat` or any other param by ref?